[FEATURE-SERVER-SIDE] Add force option

apelisse opened the pull request:
Create PatchOptions (embedding UpdateOptions) and add the force flag to it. It doesn't do anything for the moment, I just want to add it. I also haven't regenerated the code, so it's probably not going to work, but running the generation takes a long time so I want to make sure we agree on the interface/documentation before :-)

**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
NONE
```

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -518,6 +518,11 @@ type UpdateOptions struct {
+	// Force is going to "force" Apply requests. That means that the
+	// user will re-acquire conflicting fields owned by other people.
+	// Force flag is ignored for non-apply requests.
+	Force bool `json:"force,omitempty" protobuf:"bool,2,rep,name=force"`
```
Do UpdateOptions work for both PUT and PATCH?

apelisse commented:
Yes

On Thu, Nov 8, 2018 at 7:10 PM Daniel Smith <notifications@github.com>
wrote:

> *@lavalamp* commented on this pull request.
> ------------------------------
>
> In staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/70824#discussion_r232131972>
> :
>
> > @@ -518,6 +518,11 @@ type UpdateOptions struct {
>  	// - All: all dry run stages will be processed
>  	// +optional
>  	DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
> +
> +	// Force is going to "force" Apply requests. That means that the
> +	// user will re-acquire conflicting fields owned by other people.
> +	// Force flag is ignored for non-apply requests.
> +	Force bool `json:"force,omitempty" protobuf:"bool,2,rep,name=force"`
>
> Do UpdateOptions work for both PUT and PATCH?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/70824#pullrequestreview-173261110>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1xrObUI2Naw5yWoImQlEAqOP-i6I-bks5utPI9gaJpZM4YVjOZ>
> .
>

Commit:
Create PatchOptions with Force option

Create a new PatchOptions that embeds the UpdateOptions, and adds a
patch specific option Force. The force option only makes sense for the
apply content-type and thus will reject any non-apply request that
includes it.

apelisse commented:
@lavalamp Updated, PTAL
@roycaihw Do you remember what needs to be updated so that it shows in the OpenAPI? I'm afraid we're going to have the same bug we did with dryRun ...

roycaihw commented:
@apelisse You need to add `PatchOptions` object parameters to PATCH route in installer like: https://github.com/kubernetes/kubernetes/pull/69359/commits/e0162b9105080f9472c1cbbab82a90b6ac2baebc.

apelisse commented:
@roycaihw Add new commit, PTAL!

roycaihw commented:
LGTM for the installer change. Please run `hack/update-openapi-spec.sh` to re-generate the spec.

lavalamp commented:
Code changes look good, but I didn't see a test?

apelisse commented:
Added a test, if everything looks good, I'll run the generation scripts.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/apiserver_test.go:
```diff
@@ -3867,6 +3867,33 @@ func (storage *SimpleRESTStorageWithDeleteCollection) DeleteCollection(ctx conte
+	defer server.Close()
+	baseUrl := server.URL + "/" + prefix + "/" + testGroupVersion.Group + "/" + testGroupVersion.Version + "/namespaces/default/simples"
+	response := runRequest(t, baseUrl, "PATCH", []byte(`{"labels":{"foo":"bar"}}`), "application/merge-patch+json; charset=UTF-8")
+	if response.StatusCode == http.StatusBadRequest {
```
Why == BadRequest rather than != OK?

lavalamp commented:
/approve

Commit:
Run various generation scripts

apelisse commented:
/retest

roycaihw commented on staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:
```diff
@@ -691,7 +694,7 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 				Reads(metav1.Patch{}).
 				Writes(producedObject)
-			if err := addObjectParams(ws, route, versionedUpdateOptions); err != nil {
+			if err := addObjectParams(ws, route, versionedPatchOptions); err != nil {
```
The `dryRun` query parameter is removed from generated openapi spec.

/hold

I realized that `addObjectParams` only parses top-level non-struct fields. `dryRun` gets dropped because UpdateOptions is nested in PatchOptions. You may want to keep `addObjectParams(ws, route, versionedUpdateOptions)` to reflect both `dryRun` and `force` query parameters. 
https://github.com/kubernetes/kubernetes/blob/e3ddaaad1d66ce1f74d6fa809d834c4de64546f9/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L912-L914

apelisse replied:
Fixed, feel free to PTAL but as far as I can tell, the swagger looks better :-)

apelisse commented:
/retest

Commit:
Update DynamicClient Patch to use PatchOptions

Commit:
Test that PatchOptions is a superset of UpdateOptions

jennybuckley commented:
/lgtm
